src/Abstractions/TaskActivity.cs

```diff
@@ -89,31 +83,31 @@ public abstract class TaskActivity<TInput, TOutput> : ITaskActivity
     /// <param name="input">The deserialized activity input.</param>
     /// <returns>The output of the activity as a task.</returns>
     public abstract Task<TOutput> RunAsync(TaskActivityContext context, TInput input);
+}
 
-    /// <summary>
-    /// Due to nullable reference types being static analysis only, we need to do our best efforts for validating the
-    /// input type, but also give control of nullability to the implementation. It is not ideal, but we do not want to
-    /// force 'TInput?' on the RunAsync implementation.
-    /// </summary>
-    static bool IsValidInput(object? input, [NotNullWhen(true)] out TInput? typedInput)
-    {
-        if (input is TInput typed)
-        {
-            // Quick pattern check.
-            typedInput = typed;
-            return true;
-        }
-        else if (input is not null && typeof(TInput) != input.GetType())
-        {
-            typedInput = default;
-            return false;
-        }
+/// <inheritdoc cref="TaskActivity{TInput, Unit}" />
+public abstract class TaskActivity<TInput> : ITaskActivity
```

Comment:
I realized a gap with the existing `TaskActivity<TInput, TOutput>` and `TaskOrchestrator<TInput, TOutput>` in the mediator pattern: generic constraints missing. Ideally, we would have this contract:

``` CSharp
public abstract class TaskActivity<TInput, TOutput> : ITaskActivity
    where TInput : IActivityRequest<TOutput>
```

The purpose here is to force the `TOutput` of the `IActivityRequest` and the `TaskActivity` to line up, so there is no risk of mismatched type failing at runtime. This would be a breaking change to introduce though. It also _forces_ you to use request objects as inputs - I would like to retain the ability to use existing types.

I could introduce another derived type for this? But what would the incentive be to use it over the existing types?

Reply:
A new derived type makes sense. As far as incentives for using it, that's a little tricky. On one hand, the incentive of having strongly typed means of calling activities/orchestrations is really nice. On the other hand, we have source generators for that which work with the existing abstractions.

One thought I just had is that Dapr Workflow (which will be another source of customers for this SDK) has a .NET SDK that's a wrapper around this SDK. Because it's a wrapper with completely different types, we can't really take advantage of the existing source generators. It feels like it would be a waste to duplicate that effort and create Dapr Workflow source generators. But if we have the typed mediator pattern in this SDK, we could change the Dapr Workflow SDK to use this new abstraction to provide a more typed experience (that project is in alpha, so we can make breaking changes still). Happy to discuss this further if you're interested.